modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/ags_solver.cc

```diff
@@ -75,6 +96,24 @@ AGSSolver::Solve()
     lbs_solver_.WriteRestartData();
 }
 
+double
+AGSSolver::ComputePhiErrorNorm() const
+{
+  double norm = 0.0;
+  auto& phi_new = lbs_solver_.PhiNewLocal();
+  for (int i = 0; i < phi_new.size(); ++i)
+  {
+    double val = phi_new[i] - phi_old_[i];
+    norm += val * val;
+  }
```

Comment:
we do not have a norm function for vectors ?

Reply:
I'm pretty sure we do, but, doing it this way, eliminated the need to store another temporary copy of phi.

Reply:
I think so. This is a tiny bit of code, and I think we should be memory conscious - even at the cost of duplicating a bit of code here and there.

Reply:
What about writing a specific L2-error routing in math.h/cc that takes two vectors? That way we don't have to write a new L2-error routine every time we need to access a different solution vector in a class.

Reply:
After thinking about this, I don't see how it's possible to do this by invoking a serial L2-error routine on each location. Computing the 2-norm of each individual vector and adding is incorrect. The only correct way to do this is to do a global reduce on the sum of the squares of the error and then have each location do a sqrt.

Reply:
We could create a routine in `lbs_solver` that does this type of convergence checking, but I think it might be best to keep it local to `ags_solver` for now. As I have mentioned, I hope to pull these routines out of `lbs_solver` at some point, and we could refactor this functionality at that time.

Reply:
> After thinking about this, I don't see how it's possible to do this by invoking a serial L2-error routine on each location. Computing the 2-norm of each individual vector and adding is incorrect. The only correct way to do this is to do a global reduce on the sum of the squares of the error and then have each location do a sqrt.

Why couldn't that be done in a function in math.h/.cc?

> We could create a routine in lbs_solver that does this type of convergence checking, but I think it might be best to keep it local to ags_solver for now. As I have mentioned, I hope to pull these routines out of lbs_solver at some point, and we could refactor this functionality at that time.

I am fine punting things, I suppose... But we should try to minimize introducing code that we admit needs a refactor before hitting master.

Reply:
Fixed. I've added L2 and pointwise checks to `math.h/.cc`. These routines operate on `std::vector<double>`. I've also added analogous checks to `lbs_solver/iterative_algorithms/convergence.h/.cc` that operate on LBS-structured data.